feat: middleware

gkatsev opened the pull request:
## Description
This is the initial PR for middleware which hopes to solve the problems from https://github.com/videojs/video.js/issues/3734.
Middleware are objects with two special functions `setSource` and `setTech`. On top of these special methods, it can have getters and setters. Currently, the only allowed ones are `currentTime`, `duration`, and `setCurrentTime`. In the future it can be extended to allow more. Also, methods like `play` could probably be treated as setters and events originating from the video element can be treated as getters and events triggered on the player could be treated as setters. But that's more in the future.

The core of middleware is that video types are basically a routing tree. So, you can "use" a middleware on a particular type but you can also "use" a tech on a particular type using special syntax since techs need a bit extra handling.
```js
// example HLS usage
videojs.use('application/x-mpegurl', require('contrib-hls'));
videojs.use('application/vnd.apple.mpegurl', require('contrib-hls'));

// example tech usage
videojs.use('video/mp4', 'videojs/html5');
videojs.use('application/x-mpegurl', 'videojs/html5');
```

The contrib-hls object could look something like this:
```js
module.exports = {
  setSource(srcObj, next) {
    // next expects an error as a first argument to let the middleware know whether you can handle the current srcObj
    // you can also modify this srcObj and return something else instead which will continue down the middleware tree
    next(false, srcObj);
  },

  setTech(tech) {
    // this gets the tech that ends up getting chosen to back everything. Store it if you need it.
    // the tech still acts as the final source of truth for everything
  },

  currentTime(previousCt) {
    // getters get the value from further up the chain and should return a transformed value, if necessary.
    // In a simple case, the middleware would get the value from the tech and would expect to return a value for the player to use
    return previousCt / 2;
  },

  setCurrentTime(newCt) {
    // setters get the value from further up the chain and should return a transformed value, if necessary.
   // In a simple case, the middleware would get the value from the player and would expect to return a value for the tech to use
  return newCt;
  }
};
```

I have a simple live example here: http://plnkr.co/edit/Ap9Fvg21lpYW3Rnlk9ty?p=preview

gkatsev commented on package.json:
```diff
@@ -20,9 +20,15 @@
     "build": "grunt dist",
     "change": "grunt chg-add",
     "clean": "grunt clean",
+    "grunt": "grunt",
```
Most of this will get removed before it's ready. Makes working with this locally a bit easier.
It'll probably get pulled into a separate PR.

gkatsev commented on src/js/player.js:
```diff
@@ -1255,6 +1266,24 @@ class Player extends Component {
+  }
+
+  middlewareGet(method) {
+    return this.middleware_.reduceRight(this.middlewareIterator(method), undefined);
```
Finally, I can use `reduce` and `reduceRight`!

gkatsev commented on src/js/video.js:
```diff
@@ -736,19 +736,9 @@ videojs.insertContent = Dom.insertContent;
- */
-if (typeof define === 'function' && define.amd) {
-  define('videojs', [], () => videojs);
+module.exports = videojs;
```
This will get pulled into a separate PR.

boushley commented:
I haven't dug into the code yet, but from the description it seems odd that for `setSource` and `setCurrentTime` follow different patterns.

What is the purpose of the `error` argument passed to `next` in `setSource`?

gkatsev commented on src/js/video.js:
```diff
@@ -736,19 +736,9 @@ videojs.insertContent = Dom.insertContent;
-  module.exports = videojs;
-}
+videojs.use = middlewareUse;
+videojs.use('video/mp4', 'videojs/html5');
```
How to register the techs for native functionality is still something I'm a bit unsure about.

boushley replied:
Definitely something we need to think about. Because as it stands you can never add a middleware that handles `setSource` for 'video/mp4' since there is a tech registered as the first item always.

boushley commented on src/js/tech/middleware.js:
```diff
@@ -0,0 +1,63 @@
+
+      // we've succeeded, now we need to go deeper
+      acc.push(mw);
+      ssh(_src, middlewares[_src.type] || [], next, acc);
```
We need some way in this case to prevent us from calling the same middleware over and over. If the middleware leaves the `src.type` unchanged and doesn't send along an error, we will keep calling the same middleware over and over.

boushley commented on src/js/tech/middleware.js:
```diff
@@ -0,0 +1,63 @@
+
+      // something happened, try the next middleware on the current level
+      if (err) {
+        return ssh(src, middleware.slice(1), next, acc);
```
Instead of doing a `return` here which implies that we care about the return value, can we just put an `else` on the if statement and add the following two lines to it.

This will also make the code more consistent with other recursive calls in this function which don't return.

boushley commented on src/js/tech/middleware.js:
```diff
@@ -0,0 +1,63 @@
+
+    // we've reached a fork, so, we need go deeper
+    } else {
+      ssh(src, middlewares[mw] || [], next, acc);
```
So this makes it possible to do something like `videojs.use('video/foo', 'my-standard-middleware-stack')` and when that "middleware" is hit it would then start processing the `my-standard-middleware-stack` set of middlewares, right? Seems interesting.

Although assuming that set of middleware eventually changes the type back to something standard we need to determine if we will ever call the same middleware twice.

boushley commented on src/js/tech/middleware.js:
```diff
@@ -0,0 +1,63 @@
+const middlewares = {};
+
+export function use(type, middleware) {
+  (middlewares[type] = middlewares[type] || []).push(middleware);
```
I'm too new to the project to know the general code style guidelines, but I find that lines like this while clever add an unnecessary obstacle to reading the code.

I feel that dividing this up into two lines makes it more idiomatic, and thus easier to read:

```
middlewares[type] = middlewares[type] || []
middlewares[type].push(middleware)
```

boushley replied:
:D

mmcc replied:
If we're picking this nit, I generally prefer `concat` to `|| []`.

```
middlewares[type] = [].concat(middlewares[type], middlware)
```

This is a little bit of nonsense here since we're ok with mutating the original array, so ¯\\\_(ツ)_/¯

BrandonOCasey replied:
I also think that "Type" should be renamed to `mimeType`. I am also wondering if it would make more sense for middleware to come first in the arguments order.  I want to `use()` `hls` for `x` type(s). I also thing that type should be able to be an array.

misteroneill replied:
@mmcc The only problem with `concat` in that case is that if `middlewares[type]` is `undefined`, you'll get an array like `[undefined, middleware]` which I think is probably undesirable.

gkatsev replied:
I went for the two line method.

boushley commented on src/js/player.js:
```diff
@@ -1255,6 +1266,24 @@ class Player extends Component {
+  }
+
+  middlewareSet(method, arg) {
+    return this.middleware_.reduce(this.middlewareIterator(method), arg);
```
Hooray for the right tool for the job! (`reduce` here and `reduceRight` below!)

boushley commented on src/js/player.js:
```diff
@@ -1255,6 +1266,24 @@ class Player extends Component {
+
+  middlewareGet(method) {
+    return this.middleware_.reduceRight(this.middlewareIterator(method), undefined);
+  }
```
It feels like these three middleware functions might belong in the `middleware.js` file instead. Just be that they accept a list of middlewares to process that would be passed in.

gkatsev replied:
The methods should be private. The reason why they aren't in `middleware.js` and why `setSource` is special is because `setSource` traverses through our middleware tree and sets up the specific middleware to be used for the player at this time. Though, I guess it can be moved into `middleware.js` and then `this.middleware_` could be passed in.

boushley replied:
Right, I was assuming that `this.middleware_` would be passed in.

The `player.js` file is big enough as it is. Seems like iterating over a list of middlewares should be in the middlewares file instead.

gkatsev commented:
Should middleware be given a reference to the `player`?

boushley commented:
It seems like it could add some new capabilities. In fact to implement the use case I talked about before (unmute sets volume to above 0) you would need to handle the mute setter and be able to set volume on the player.

This could also cause problems if someone does something silly like tries to set `currentTime` on the player while in a middleware that is handling `currentTime`.

mmcc commented on src/js/player.js:
```diff
@@ -1843,6 +1872,31 @@ class Player extends Component {
+      this.middleware_ = mws;
+      this.loadTech_(tech.name, src_);
+      middleware.setTech(mws, this.tech_);
+    });
```
If `tech.name` is the only thing used here, we could destructure in the args. I don't feel strongly about this one, just something I've started doing a little more of lately.

```
middleware.setSource(src, ({ name }, src_, mws) => {
  this.middleware_ = mws;
  this.loadTech_(name, src_);
  middleware.setTech(mws, this.tech_);
});
```

mmcc commented on src/js/tech/middleware.js:
```diff
@@ -0,0 +1,70 @@
+}
+
+export function setSource(src, next) {
+  setTimeout(()=>ssh(src, middlewares[src.type] || [], next, []), 1);
```
I thought Standard required spaces here (i.e. `() => foo()`)?

I don't feel crazy strong either way, but on [line 16](https://github.com/videojs/video.js/pull/3788/files#diff-b1bec4e9f074fe980f0471fcaaf4df75R16) there are spaces so we're not even being consistent within the same file.

gkatsev replied:
Apparently not but yeah, I agree that there should be more whitespace. Originally added it in to verify that async works.

BrandonOCasey commented on src/js/player.js:
```diff
@@ -2379,6 +2393,91 @@ class Player extends Component {
+      return true;
+    }
+
+    return;
```
should we return false here to keep the return value consistent?

gkatsev replied:
makes sense.

BrandonOCasey commented on src/js/player.js:
```diff
@@ -2379,6 +2393,91 @@ class Player extends Component {
+  src_(source) {
+    const sourceTech = this.selectSource([source]);
+
+    if (sourceTech) {
```
seems like this should be 
```js
if (!sourceTech) {
  return true;
}
```

gkatsev replied:
yeah, makes it less nested.

BrandonOCasey commented on src/js/player.js:
```diff
@@ -2379,6 +2393,91 @@ class Player extends Component {
+    const sourceTech = this.selectSource([source]);
+
+    if (sourceTech) {
+      if (sourceTech.tech === this.techName_) {
```
It would probably be good to have a similar shortout here. 

```js 
if (sourceTech.tech !== this.techName_) {
  // load this technology with the chosen source
  his.loadTech_(sourceTech.tech, sourceTech.source);
  return false;
}
```

BrandonOCasey commented on src/js/player.js:
```diff
@@ -2379,6 +2393,91 @@ class Player extends Component {
+          if (this.tech_.constructor.prototype.hasOwnProperty('setSource')) {
+            this.techCall_('setSource', source);
+          } else {
+            this.techCall_('src', source.src);
```
Should we deprecate the old way?

gkatsev replied:
I don't think we're quite ready to do that yet.

BrandonOCasey commented on src/js/player.js:
```diff
@@ -1530,7 +1535,12 @@ class Player extends Component {
+    if (method in middleware.allowedSetters) {
+      return middleware.set(this.middleware_, this.tech_, method, arg);
+    }
+
```
seems like we should check do this here, to prevent checking `this.tech_` in two places below.

```js
if (!this.tech_) {
  return
}  
```

Also it seems like `this.tech_.ready()` already checks if the tech is ready? Seems like we could merge a lot of this code together to make it much simpler.

gkatsev replied:
You're right, we probably want to put the middleware section into one of the ifs.
It's possible that just doing .ready() works, in which case, I can see if it's worth trying to simplify it for this PR.

BrandonOCasey commented on src/js/player.js:
```diff
@@ -1562,6 +1572,10 @@ class Player extends Component {
   techGet_(method) {
     if (this.tech_ && this.tech_.isReady_) {
 
+      if (method in middleware.allowedGetters) {
```
This function could also use some similar cleanup, maybe we can do both in another PR, but its always bothered me that it seems so messy an nested.

gkatsev replied:
Same here. I'd prefer to do the cleanup in a separate PR but might do it if it helps make the middleware part more clear as well.

gkatsev replied:
we can't actually do this cleanup here because we need to return a value from this function and `this.ready()` doesn't guarantee a value, even if you're running it synchronously. However, I think I was able to simplify `techCall`.

BrandonOCasey commented on src/js/tech/middleware.js:
```diff
@@ -0,0 +1,79 @@
+}
+
+export function setSource(setTimeout, src, next) {
+  setTimeout(() => ssh(src, middlewares[src.type], next), 1);
```
will we get this far if a `src` does not have a `type`? Should we be checking for this somewhere?

BrandonOCasey replied:
should this be async?

BrandonOCasey commented on src/js/tech/middleware.js:
```diff
@@ -0,0 +1,79 @@
+  setTimeout(() => ssh(src, middlewares[src.type], next), 1);
+}
+
+export function setTech(middleware, tech) {
```
middlewareList since it seems like we want it to be an array here?

BrandonOCasey commented on test/index.html:
```diff
@@ -9,6 +9,9 @@
   <div id="qunit"></div>
   <script src="../node_modules/qunitjs/qunit/qunit.js"></script>
+  <script>
+    QUnit.config.reorder = false;
```
Should this be in the generator/spellbook?

misteroneill replied:
I don't think so. I think setting `reorder` to `false` is really used to mask bad cleanup in tests. I think @gkatsev may have cleaned that up in another PR as well.

gkatsev replied:
I rebased against master, so, I'll try again without it.

gkatsev replied:
Yep, it can be removed now.

misteroneill commented on src/js/tech/middleware.js:
```diff
@@ -0,0 +1,79 @@
+  };
+}
+
+function ssh(src = {}, middleware = [], next, acc = []) {
```
Agreed. I keep thinking of SSH.

gkatsev replied:
Yep, the name should definitely change.

misteroneill commented on src/js/video.js:
```diff
@@ -767,19 +768,6 @@ videojs.insertContent = Dom.insertContent;
+module.exports = videojs;
 
-export default videojs;
+videojs.use = middlewareUse;
```
There a reason this comes after `module.exports`?

gkatsev replied:
Nope.

misteroneill commented on test/unit/player.test.js:
```diff
@@ -283,6 +281,7 @@ QUnit.test('should asynchronously fire error events during source selection', fu
   });
 
   this.clock.tick(1);
+  this.clock.tick(1);
```
Could we add an explanation as to why we need to tick 2ms instead of 1ms?

gkatsev replied:
It's because the error is triggered async, so, we need one for it, but we need an additional one for the player creation itself.

gkatsev replied:
I added a comment

gkatsev commented on package.json:
```diff
@@ -25,7 +25,13 @@
+    "predev": "babel src/js -d es5",
+    "dev": "run-p babel browserify",
+    "dev-test": "run-p dev test-watch",
+    "babel": "babel -w src/js -d es5",
```
I'm going to pull these out of this PR when getting ready to merge and make a separate PR out of it.

gkatsev commented on src/js/player.js:
```diff
@@ -829,6 +834,8 @@ class Player extends Component {
       'videoTracks': this.videoTracks_,
       'textTracks': this.textTracks_,
+      'remoteTextTracks': this.remoteTextTracks_,
+      'remoteTextTrackEls_': this.remoteTextTrackEls_,
```
Unfortunately, this change does need to be in this PR, some of the tests rely on tracks and will fail without it because techs are now async.

gkatsev commented on src/js/video.js:
```diff
@@ -767,19 +768,6 @@ videojs.insertContent = Dom.insertContent;
- * still support requirejs and browserify. This also needs to be closure
- * compiler compatible, so string keys are used.
- */
-if (typeof define === 'function' && define.amd) {
```
Yeah, I'll pull it out in favor of that. I think I did this because I thought something was broken with something I tried.

gkatsev commented on test/unit/player.test.js:
```diff
@@ -596,6 +595,9 @@ QUnit.test('make sure that controls listeners do not get added too many times',
+/*
+ * sourceOrder is no longer used with middleware
+ *
 QUnit.test('should select the proper tech based on the the sourceOrder option', function(assert) {
```
I'll remove it right before merging.

gkatsev commented:
I have a question about what `.src` and friends should return. Right now it looks like this:
```html
  <video width=600 height=300 lang="es" id="vid1" class="video-js vjs-default-skin" controls >

      <source src="foo" type="video/foo">
    <source 
      src="//d2zihajmogu5jn.cloudfront.net/bipbop-advanced/bipbop_16x9_variant.m3u8"
      type="application/x-mpegurl"
      >
    <source src="http://vjs.zencdn.net/v/oceans.mp4" type='video/mp4'>
    <source src="http://vjs.zencdn.net/v/oceans.webm" type='video/webm'>
    <source src="http://vjs.zencdn.net/v/oceans.ogv" type='video/ogg'>

    <track kind="captions" src="text.vtt" srclang="en" label="English" default></track>

  </video>
```
And the output of the methods is this:
```js
player.src()
// "blob:http://localhost:8000/a4360e78-1eec-4f58-9be5-3e0f12cab696"
player.currentSrc()
// "//d2zihajmogu5jn.cloudfront.net/sintel/master.m3u8"
player.currentSource()
// {type: "video/foo", src: "foo"}
player.currentSources()
/* [
  {
    "type": "video/foo",
    "src": "foo"
  },
  {
    "type": "application/x-mpegurl",
    "src": "https://s3.amazonaws.com/_bc_dml/example-content/bipbop-advanced/bipbop_16x9_variant.m3u8"
  },
  {
    "type": "video/mp4",
    "src": "http://vjs.zencdn.net/v/oceans.mp4"
  },
  {
    "type": "video/webm",
    "src": "http://vjs.zencdn.net/v/oceans.webm"
  },
  {
    "type": "video/ogg",
    "src": "http://vjs.zencdn.net/v/oceans.ogv"
  }
] */
```

dmlap commented:
@gkatsev I don't think `player.src()` should return a blob URL unless the app developer (i.e. not the middleware developer) set a blob URL as the source. That was one of the big design goals for source handlers and I think it's still a good idea.

gkatsev commented:
@dmlap That is the current behavior with source handlers. Also, historically, we've made the methods like `.src()` just match what the video element returns. We've definitely already diverged a bit with regards to `currentSrc()` in this case. Now might be the time to diverge a bit more if we feel that's the right direction.
I kind of like that we have a way to get exactly the src from the video element.

Also, I should add that in the above scenario the `foo` source is the one that was selected.

misteroneill commented:
I think it makes sense to keep things as similar to the current state as possible. Or adhering to the standard (in the case of `src()` and `currentSrc()`).

As an aside, the `blob:` URL is a consequence of HLS, right?

gkatsev commented:
After some discussion with @dmlap, we think that `currentSrc()` should regular the actual source that is being set, like `foo`. `src()` should probably return `foo` as well in this case but maybe it should instead return the `m3u8` source. Also, `currentType()` should match the type of the source in `currentSrc()`.

misteroneill commented:
So, if it's returning the source that was originally set (pre-middleware activity), is there going to be a way to get the post-middleware source?

gkatsev commented:
I think it would then be up to the middleware and potential corresponding plugin to surface. Plus, there's always asking the tech directly.

gkatsev commented:
I've moved the grunt and package.json changes into a separate PR: https://github.com/videojs/video.js/pull/3896 and thus I reverted the changes in [this PR](https://github.com/videojs/video.js/pull/3788/commits/7429e65ab712ecbdccebb4acc9b2145bf470a006).

gkatsev commented:
@mmcc @boushley @misteroneill I've changed the PR from WIP to RFC. I think it's in decent shape now just needs closer inspection.

BrandonOCasey commented:
@gkatsev It would be a good idea to run the jsdoc linter over this code

misteroneill commented on src/js/tech/middleware.js:
```diff
@@ -0,0 +1,88 @@
+    setSourceHelper(src, mwrest, next, acc);
+  } else {
+    // const techs = [].concat(middlewares['*']);
+    // const tech = techs.filter((tech) => !!tech.tech.canPlayType(src.type))[0].name;
```
Should these commented lines just be removed?
